sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -230,9 +231,19 @@ case class Union(children: Seq[LogicalPlan]) extends LogicalPlan {
   }
 
   // updating nullability to make all the children consistent
-  override def output: Seq[Attribute] =
-    children.map(_.output).transpose.map(attrs =>
-      attrs.head.withNullability(attrs.exists(_.nullable)))
+  override def output: Seq[Attribute] = {
+    children.map(_.output).transpose.map { attrs =>
+      val firstAttr = attrs.head
+      val outAttr = try {
+        firstAttr.withDataType(attrs.map(_.dataType).reduce(StructType.merge))
+      } catch {
+        // If the data types are not compatible (eg. Decimals with different precision/scale)
+        // return the first type
```

Comment:
All type compatibly checks should be in `resolved`?

Reply:
yes, right. Let me update the comment. Thanks.